### PR TITLE
KTOR-4519 Add Debug Logging to Auth Plugin

### PR DESCRIPTION
**Pull request by e5l**
Fix [KTOR-4519](https://youtrack.jetbrains.com/issue/KTOR-4519/Add-Debug-Logging-to-Auth-Plugin)

### CODE REVIEW

File: ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt [hunk 1]
```diff
@@ -55,88 +58,90 @@ public object AuthenticationChecked : Hook<suspend (ApplicationCall) -> Unit> {
...
+                if (configurationName == null) {
+                    "Default authentication configuration was not found"
+                } else {
+                    "Authentication configuration with the name $configurationName was not found"
+                }
+            )
+        }
+        for (provider in configurations) {
+            if (provider.skipWhen.any { skipCondition -> skipCondition(call) }) continue
+
```
Comment (rsinukov) [inline ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt:85]: maybe add here a log `Authenticating ${call.request.uri} with provider $provider`?

File: ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt [hunk 2]
```diff
@@ -55,88 +58,90 @@ public object AuthenticationChecked : Hook<suspend (ApplicationCall) -> Unit> {
...
-                if (authenticationContext.principal != null) break
+        if (context.principal != null) return@on
+        if (context.challenge.completed) {
+            if (!call.isHandled) {
+                call.respond(UnauthorizedResponse())
             }
+            return@on
+        }
+        if (pluginConfig.optional && context.allFailures.none { it == AuthenticationFailedCause.InvalidCredentials }) {
+            return@on
```
Comment (rsinukov) [inline ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt:106]: please add a log here

File: ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt [hunk 3]
```diff
@@ -55,88 +58,90 @@ public object AuthenticationChecked : Hook<suspend (ApplicationCall) -> Unit> {
...
         }
 
-        on(ChallengeHook) { call ->
-            val context = AuthenticationContext.from(call)
+        val challenges = context.challenge.challenges
 
-            if (context.principal != null) return@on
+        for (challenge in challenges) {
+            challenge(context.challenge, call)
```
Comment (rsinukov) [inline ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt:112]: should we add logs inside providers and challenges?

**Review by rsinukov — CHANGES_REQUESTED** [review]

File: ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt (hunk 1 above)
Comment (e5l) [reply ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt:85]: Fixed

File: ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt (hunk 2 above)
Comment (e5l) [reply ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt:106]: Thanks, done!

File: ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt (hunk 3 above)
Comment (e5l) [reply ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt:112]: let's make it as a separate task

File: ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt (hunk 1 above)
Comment (rsinukov) [reply ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt:88]: I still don't see a log here =)

**Review by rsinukov — APPROVED** [review]
lgtm, but 1 comment

File: ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt (hunk 1 above)
Comment (e5l) [reply ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt:88]: sorry, now fixed

**Pull request merged**